zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1165,20 +1177,92 @@ private void reconnect() {
     }
   }
 
+  private class SyncContext {
+    private CountDownLatch _latch;
+    private AtomicInteger _rc;
+
+    public SyncContext(CountDownLatch latch) {
+      _latch = latch;
+    }
+
+    AtomicInteger getRc() {
+      return _rc;
+    }
+
+    void setRc(AtomicInteger rc) {
+      _rc = rc;
+    }
+
+    CountDownLatch getLatch() {
+      return _latch;
+    }
+  }
+
+  private boolean retrySync(String sessionId) throws ZkInterruptedException {
+    if (!_syncOnNewSession) {
+      return true;
+    }
+    while (true) {
+      CountDownLatch latch = new CountDownLatch(1);
+      SyncContext ctx = new SyncContext(latch);
+
+      final ZkConnection zkConnection = (ZkConnection) getConnection();
+      zkConnection.getZookeeper().sync(_syncPath, new AsyncCallback.VoidCallback() {
+        @Override
+        public void processResult(int rt, String s, Object ctx) {
+          LOG.info("sycnOnNewSession with sessionID {} async return code: {}", sessionId, rt);
+          SyncContext synCtx = ((SyncContext) ctx);
+          synCtx.setRc(new AtomicInteger(rt));
+          synCtx.getLatch().countDown();
+        }
+      }, ctx);
+
+      try {
+        latch.await();
+      } catch (InterruptedException e) {
+        LOG.info("retrySync latch waiting got interrrupted with sessionId {} ", sessionId);
+        throw new ZkInterruptedException(e);
+      }
+
+      KeeperException.Code code = KeeperException.Code.get(ctx.getRc().get());
+      if (code == OK) {
+        LOG.info("sycnOnNewSession with sessionID {} async return code: {} and proceeds", sessionId,
+            code);
+        break;
+      }
+      if (code == CONNECTIONLOSS || code == SESSIONMOVED) {
```

Comment:
Except these three types of code, is there any other code need to be handled?

Reply:
We intentionally not to handle others, including session expiration. See comment on line 1238